pkg/allocator/allocator.go

```diff
@@ -891,16 +891,66 @@ type RemoteCache struct {
 // kvstore will be maintained in the RemoteCache structure returned and will
 // start being reported in the identities returned by the ForeachCache()
 // function. RemoteName should be unique per logical "remote".
-func (a *Allocator) WatchRemoteKVStore(remoteName string, remoteAlloc *Allocator) *RemoteCache {
+func (a *Allocator) WatchRemoteKVStore(ctx context.Context, remoteName string, remoteAlloc *Allocator) (*RemoteCache, error) {
 	rc := &RemoteCache{
 		cache: &remoteAlloc.mainCache,
 	}
 
+	scopedLog := log.WithField(fieldRemote, remoteName)
+	scopedLog.Info("Waiting for remote KVStore watcher synchronization")
+
+	select {
+	case <-ctx.Done():
+		scopedLog.Warning("Context canceled before remote KVStore watcher synchronization completed")
+		rc.Close()
+
+		a.remoteCachesMutex.Lock()
+		old := a.remoteCaches[remoteName]
+		a.remoteCachesMutex.Unlock()
+
+		if old != nil {
+			old.cache.mutex.RLock()
+			defer old.cache.mutex.RUnlock()
+		}
+
+		// Drain all entries that might have been received until now, and that
+		// are not present in the current cache (if any). This ensures we do not
+		// leak any stale identity, and at the same time we do not invalidate the
+		// current state.
+		rc.cache.drainIf(func(id idpool.ID) bool {
```

Comment:
It doesn't, because `MasterKeyProtection` is not enabled for remote kvstores (as they are read-only) The only effective action is `c.sendEvent(kvstore.EventTypeDelete, id, key)`, but I preferred to directly call `onDeleteLocked` for consistency with standard operations (let me know if you prefer I switch to that).